### DISCUSSION TITLE
getStaticPaths & MDX

### DISCUSSION THREAD

**Discussion by guidovizoso** [vercel/next.js#12322] [Help]
Hi everyone!

I'm trying to build a website that uses Next and MDX. To separate the urls from the file names I'm using getStaticPaths and getStaticProps to build the slugs from front matter with gray-matter.

The problem is the returned MDX from getStaticProps is a function and I can't return a funciton from getStaticProps. If I serialize the content then I'm not able to use imports.

Is there a way to overcome this or am I over complicating the solution? I'll share the code on my [...slug].js file in case it's helpful.

Thanks in advance!

```
import fs from 'fs';
import path from 'path';
import matter from 'gray-matter';
import { transpileMarkdown } from '../../utils/mdx';

function Post({ data, content }) {
  // At this point content is a string.
  return <div>{JSON.stringify({ data })}</div>;
}

export async function getStaticPaths() {
  const postDir = await path.join(process.cwd(), 'posts');
  const postList = await fs.readdirSync(postDir);
  let pathList = await postList.map((p) => {
    let post = require('../../posts/' + p).default;
    const { data } = matter(post);
    return { params: { slug: [data.category, data.slug] } };
  });
  return {
    paths: pathList,
    fallback: false,
  };
}

export async function getStaticProps({ params }) {
  const { slug } = params;
  const postDir = await path.join(process.cwd(), 'posts');
  const postList = await fs.readdirSync(postDir);
  let foundPost = postList.reduce(function (acc, p) {
    let post = require('../../posts/' + p).default;
    const { content, data } = matter(post);
    if (slug[0] === data.category && slug[1] === data.slug)
      return { content, data };
  }, null);
  let transpiledPost = await transpileMarkdown({ md: foundPost.content });
  let serializedPost = await transpiledPost.toString();
  return { props: { data: foundPost.data, content: serializedPost } };
}
export default Post;
```

**Comment by valse** [comment]
With the transpiled MDX content you must wrap the resulting function with a [MDXRenderer](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-plugin-mdx/mdx-renderer.js) like component and then use the MDXProvider to specify the custom components used in the MDX content:
```jsx
<MDXProvider components={{ Button: MyCustomButton }}>
   <MDXRenderer>{content}</MDXRenderer>
</MDXProvider>

**Comment by leerob** [comment]
Do you need to use the SSG methods when you're dealing with MDX files locally? Personally, I don't see any advantage of having a `posts` folder with [x] number of Markdown files and a single dynamic route. Why not move `posts` directly to `pages` and configure your application to load `.mdx` routes? 

I can see the former approach if you're pulling Markdown from an external data source. But when you're dealing with your local file system, it seems easier to put everything in pages. 

Example: https://github.com/leerob/leerob.io/tree/master/pages/blog

**Reply by guidovizoso** [reply]
Hi @leerob! I'm taking this approach because I'd like to generate my urls dynamically. If I move everything to the `pages` directory then I'll have to create a folder for each category, making it exponentially unorganised. I know this is not exactly a huge problem but I'm just trying to find a new solution to a problem.

Thanks for replying

**Reply by TheThirdRace** [reply]
Hi @leerob 

I'm also looking into generating them dynamically at build time because I could separate the UI and Data concerns.

The way I see it...

I create posts in markdown and focus on content only, with a side of front-matter. This is raw data, I don't have to focus on a particular tool, in fact I'm pretty much tool agnostic at this point, I write simple markdown and I'm done.

Then, I use the `/page/posts/[slug].tsx` to manage my template, add a MDXProvider to replace the markdown with my own components and sprinkle my "business" logic. It fits right in with CSS-in-JS too.

That way:
- I can reuse the same posts with multiple different templates and/or multiple routes
- I can switch components however I want (H1, H2, P, etc.) per route or even per post type (using front-matter metadata)
- I can create different experiences to target an audience in particular, for example having a `/science` and a `/science-for-kid` route. The first route would contain all the posts and have a more serious design, while the latter route for kids would only present the articles tagged for kids with different animations, different colors (CSS-in-JS fits right in here), a much more funny design altogether.

**All that while keeping the raw content just content.** Thus I separate the `consumer of data` from the `data` itself.

Putting all your markdown posts under `/page` locks you into 1 design only, not to mention the copy paste in each and every mdx file for the layout and such, the data being "dirtied" by the UI stuff, etc. Furthermore, I would argue most people won't abstract their design well enough so a change in design means modifying all your mdx files and running the risk to forget one, which lead to discrepancies... I think doing it this way is very fast, but in the long run it's actually a recipe for disaster.

**Reply by leerob** [reply]
Here's two libraries to check out that might solve your problem.

- [next-mdx-enhanced](https://github.com/hashicorp/next-mdx-enhanced)
- [next-mdx-remote](https://github.com/hashicorp/next-mdx-remote)

**Reply by TheThirdRace** [reply]
@leerob 
Thanks for the links, I'm definitely gonna check them out. The second one seems to be what I'm looking for, but I haven't tested yet so I won't scream victory yet ;)

As for your previous answers on this thread, I did peak at what you did for your website. Your github repository was very well structured and all of it made sense to me. While I still think the data shouldn't dictate the layout for the sake of separating concerns, the way you did integrate MDX is by far the easiest and the best I've seen yet.

I've just read your blog post on rewriting your website and the technologies you used. I've been doing a lot of research to rewrite mine and my whole tech stack was written in your blog post! I guess if someone else settles with the same exact tech stack than I do, I might not be that far off ;)

Thanks again for the prompt reply, I've been battling a lot with MDX in the past few days and you might have provided me the solution I was looking for. It's really appreciated

**Reply by TheThirdRace** [reply]
@leerob 
Just a follow up.

`next-mdx-remote` did solve my need and I can now generate pages from a `/page/[slug].tsx`.

It wasn't as straightforward as I would have liked, but it was entirely _**my**_ fault. I chose to use Typescript and `next-mdx-remote` comes with no examples and no types definition. Took a little while to get around that... I also tried to make it work with Preact to save a few KB in bundle size, but I kept getting errors about internal React errors. I guess `next-mdx-remote` isn't Preact compatible yet. As soon as I switched to React, everything started to work.

I did some checks with your website repo and I started to integrate your method for having `mdx` files directly under `/pages`. I'm not finished yet, but everything looks fine there so far.

I figure there's no point limiting myself to only 1 method. If I can mix and match them according to my needs, I guess it's all for the better!

Thanks again for your input, it was invaluable in sorting out how to implement this. As a bit of a newbie on NodeJs configuration and related stuff, it's always awesome to be pointed in the right direction.

**Reply by TheThirdRace** [reply]
For reference in case others run into similar issues...

**Preact + `@next/mdx` + `@mdx-js/loader` + `@mdx-js/react`**
I was able to make it work with **Preact** after all. Turns out `@mdx-js/preact` simply doesn't work with the **MDXProvider**, I had to use the **React** version instead (`@mdx-js/react`).

The only downside with this solution was the consistent slower results in Lighthouse metrics than with `markdown-to-jsx`. The difference was around 400 ms on mobile, no discernible difference on desktop. Still more than acceptable, but since I'm using **Preact** to get as fast as possible, there's no point in choosing a slower solution when both are equivalent.

**next-mdx-remote**
I did try `next-mdx-remote` and it worked great with **React**, but not **Preact**. Switching from **Preact** to **React** added around 30KB to the bundle, which is acceptable if I had no other alternative. Again, this solution was slower than Preact + `markdown-to-jsx` by a small margin, around 300 ms.

**Reply by tconroy** [reply]
hey @TheThirdRace , sounds like you've accomplished what I'm attempting. Do you have a repo / minimal reproduction you could share? Would be much appreciated!

**Reply by TheThirdRace** [reply]
@tconroy 
I actually don't have the repo on public at the moment. I'm toying around making it public, but there are parts I want to keep private and it's not as easy as I'd like.

Could you be more specific about what you're attempting because I tested so many things and I'd need a pointer... I kept notes about the different solutions tested, so I can share that easily.
